[develop] Adding "pkg.downloaded" state and support for installing patches/erratas

meaksh opened the pull request:
As discussed, this PR brings #40195 to `develop` branch, as this is the right target for the PR.

### What does this PR do?
This PR adds some improvements to Zypper and Yum execution module to support patches/errata installation and `downloadonly` parameter as well as new `pkg.downloaded` state. 

1. Allowing `downloadonly=True` on a `pkg.installed` state:
```
vim:
  pkg.installed:
    - downloadonly: True
```
2. Adds new "pkg.downloaded" state function which is a wrapper of `pkg.installed` with `downloadonly=True`:
```
vim:
  pkg.downloaded: []
```
3. Adding support for installing patches/erratas using `zypper` or `yum` module.
```
$ salt 'minion' pkg.install patches='["SUSE-SLE-SERVER-12-SP1-2016-942", "SUSE-SLE-SERVER-12-SP1-2016-1048"]'
minion:
    ----------
    dhcp:
        ----------
        new:
            4.3.3-9.1
        old:
            4.3.3-2.2
    dhcp-client:
        ----------
        new:
            4.3.3-9.1
        old:
            4.3.3-2.2
    ----------
    dmidecode:
        ----------
        new:
            2.12-7.1
        old:
            2.12-5.20
```
4. Support patches/erratas with `pkg.installed` and `pkg.downloaded`:
```
my-suse-patches:
  pkg.installed:
    - patches:
      - SUSE-SLE-SERVER-12-SP1-2016-942
      - SUSE-SLE-SERVER-12-SP1-2016-1048
```

### Tests written?

Yes

meaksh committed:
Adding support for installing patches in yum/dnf execution module

meaksh committed:
Adding support for installing patches in Zypper module

meaksh committed:
Adding unit tests for Zypper execution module

salt-jenkins commented:
@meaksh, thanks for your PR! By analyzing the history of the files in this pull request, we identified @terminalmage, @steverweber and @garethgreenaway to be potential reviewers.

damon-atkins commented:
I think I would rather see something more generic, which allows extra options to be added to a command specific to package management provider.   Rather than trying to capture all the options for package manager.    However or hopefully, download is common across all of the package managers.


Not sure of the value of a state file having a download option. Because you then need to alter the sls file again to remove the download is True.

You can run the execution module within an sls with the download and get the same result.

cachedout commented:
@meaksh Did you see the above feedback from @damon-atkins ?

meaksh commented:
Sorry for the late response here and thank you @damon-atkins for you comment!

I'm agree with @damon-atkins that allowing extra options would be a better choice in order to prevent from growing parameter to infinite. I would take a look to it during tomorrow.

IMHO having state files where you can define downloaded states makes sense and they're helpful, i.a., when you want to prepare your system for patching. Even more sense if you're dynamically generating those states.

You're right, calling an execution module within an sls would also produce the same result, but you would need a more complex yaml to define the same state, and at the same time, you lose some "declarative" style in favor of a more "imperative" state definition.

Others opinions here are more than welcome :smile:

meaksh commented:
@damon-atkins - Now those platform dependent `downloadonly` and `patches` parameters are passed as kwargs to the `pkg.installed` state function. So these two will be part of kwargs when state function is calling the corresponding package manager execution module.

The `pkg.installed` function needed some fixes to get a right state response in case any of those new parameters is present, as when `patches` are passed there's no need of `name` or `pkgs`. In case of `downloadonly`, the state should returns True even if no packages were changed.

Do you think this has now a better shape?

damon-atkins commented:
Thanks for the changes.  Keep in mine typical use case when it comes to states.  
e.g.
```
mine.send:
  module.run:
    - func: network.ip_addrs
    - kwargs:
        interface: eth0
```
If their are no smarts in a state function then it may was well be written as above.  Which results in the action being performed every time.    If the state function says time (e.g. checks the current state first) , in some shape or form then its more worth doing.

meaksh commented:
Well, at this point, due this PR enables multiple stuff, I would like to get some clear comments here about what do you think or expect from this PR, because we're now kind of blocked waiting to take some design decisions which depends on the changes included on this PR:

- Are these `pkg.downloaded` and `pkg.installed` with `downloadonly=True` states going to be accepted at some point (maybe after some refactoring) or do you think these finally makes no sense in Salt in favor of running the execution module inside the state file with `module.run`?

- Are you ok enabling `downloadonly` support for `yum` execution module?

- What about the patches installation support for `yum` and `zypper` execution module?

We always try to not diverge our Salt package from upstream Salt, so it would be much appreciated if we can get some clear feedback here in order to take a decision :smile:

Maybe we can split this PR to get discussed points in, and then another PR with pending discussion.

@cachedout @damon-atkins - What do you think? Thanks!

damon-atkins commented:
> What would be better for salt instead of salt '*' state.apply test=True to be able to do salt '*' state.apply prepare=True i.e. don't change anything, but perform any preparation needed e.g. download and test

The above is more like a wish. It would take someone in the Salt team to run with it and would be a separate PR.

It's up to the salt team to decide if pkg state should have the option of downloadonly=True or not.  I see no issues with it being in an execution module.

Don't you need to add a list_patches to the execution modules, so you can determine in the state if the patch has been installed or not? And therefore determine if a change is required.

damon-atkins commented:
Ping @terminalmage may have some thoughts.

terminalmage reviewed (changes requested):
Why include any of this in the ``pkg.installed`` state? This completely breaks idempotence in that state. Not only that, it doesn't seem to have anything to do with ``pkg.installed`` in the first place, as that state deals with packages that are *installed*, and uses ``pkg.list_pkgs`` to figure out whether or not the desired packages are installed. Also, ``patches`` is a poor name for an argument that specifies packages which should be downloaded, because it suggests that any package we want to prefetch is a patch. That is not guaranteed.

I can't recommend merging this without some significant refactoring. Here are some things that would be needed:

1. Revert all changes in pkg_resource.py and the ``pkg.installed`` state.
2. Write a function to list which packages are downloaded (possible name for the function: ``list_downloaded``). This would be like ``pkg.list_pkgs`` but would list the packages that are downloaded, instead of the packages which are installed.
3. Write a proper state function for ``pkg.downloaded``. This state function would check the downloaded packages provided by ``pkg.list_downloaded`` and download only those packages which need to be downloaded.

damon-atkins commented:
It looks like vendors have a label for packages which fix a CVE for example, I assume a label represents one or more packages.  I suspect its no different to `salt.states.pkg.group_installed`.   Not sure how group_installed currently checks pkgs are installed or not. 

I see issue with pkg.patch_install  pkg.patch_list (or some other name other than patch) as it appears to be the same as pkg.group_installed  (or group install has an option to indicate the group type like "security")

@meaksh I suggest you provide some examples of the actual yum and zypper commands in action.

meaksh commented:
@terminalmage @damon-atkins - Thank you very much for your feedback I'll try to address it ASAP.

meaksh committed:
Adding list_downloaded function to Zypper module

meaksh committed:
Adding list_downloaded function to Yum module

meaksh committed:
Adding new pkg.downloaded state

meaksh committed:
Adding documentation for pkg.downloaded state

meaksh committed:
Adding pkg.patched and pkg.patch_downloaded states

meaksh committed:
Check targets for advisory patches installation

meaksh committed:
Adds support for listing advisory patches with Zypper

meaksh committed:
Adds support for listing advisory patches with Yum

meaksh committed:
Improving function naming

meaksh committed:
Moving advisory ids checks into pkg_resource.parse_targets

meaksh committed:
Fixes _find_download_targets to call _preflight_check

meaksh committed:
Fixes parse_targets when advisory id is passed as name

meaksh committed:
Pylint fixes

meaksh committed:
Enables pkg.downloaded to verify packages after package manager call.

meaksh committed:
Adding missing kwargs parameters to pkg.install call

meaksh commented:
@terminalmage - I've reverted my previous changes and updated this PR taking care of your feedback in order to produce a better PR which can be accepted. These are the new changes made by this PR:

Zypper and Yum execution modules:
1. `pkg.list_downloaded`
2. `pkg.list_patches`
3. `pkg.list_installed_patches`
4. Adds errata/patches support on `pkg.install`. Target patches are set via new `advisory_ids` kwarg.
5. Fixes `pkg.install` to be able to detect changes on downloaded packages when using `downloadonly=True`.

New states:
1. `pkg.downloaded`
2. `pkg.patch_installed`
3. `pkg.patch_downloaded`

I've added a minor change into `pkg_resourse.py` to be able to handle `advisory` targets.

Example of use in sls file:
```
test-patch-downloaded:
  pkg.patch_downloaded:
    - advisory_ids:
      - SUSE-SLE-SERVER-12-SP2-2017-185
      - SUSE-SLE-SERVER-12-SP2-2017-487

test-patch-installed:
  pkg.patch_installed:
    - advisory_ids:
      - SUSE-SLE-SERVER-12-SP2-2017-9

test-patch-already-installed:
  pkg.patch_installed:
    - advisory_ids:
      - SUSE-SLE-SERVER-12-SP2-2017-9

test-unexisting-patch:
  pkg.patch_installed:
    - advisory_ids:
      - SUSE-SLE-SERVER-12-SP2-2017-XXX

zip:
  pkg.downloaded

test-pkg-downloaded:
  pkg.downloaded:
    - pkgs:
      - whois
```

Output from applying sls file:
```
local:
----------
          ID: test-patch-downloaded
    Function: pkg.patch_downloaded
      Result: True
     Comment: 
     Started: 18:01:51.482207
    Duration: 37572.555 ms
     Changes:   
              ----------
              kernel-default:
                  ----------
                  new:
                      ----------
                      4.4.49-92.14.1:
                          /var/cache/zypp/packages/SLE-12-SP2-x86_64-Update/x86_64/kernel-default-4.4.49-92.14.1.x86_64.rpm
                  old:
              kernel-firmware:
                  ----------
                  new:
                      ----------
                      20160516git-19.4:
                          /var/cache/zypp/packages/SLE-12-SP2-x86_64-Update/noarch/kernel-firmware-20160516git-19.4.noarch.rpm
                  old:
----------
          ID: test-patch-installed
    Function: pkg.patch_installed
      Result: True
     Comment: 
     Started: 18:18:51.491957
    Duration: 5211.851 ms
     Changes:   
              ----------
              libsolv-tools:
                  ----------
                  new:
                      0.6.26-2.27.3.3
                  old:
                      0.6.23-2.23.4
              libzypp:
                  ----------
                  new:
                      16.5.2-27.9.1
                  old:
                      16.2.4-19.5
              lsof:
                  ----------
                  new:
                      4.89-25.1
                  old:
              zypper:
                  ----------
                  new:
                      1.13.21-18.3.3
                  old:
                      1.13.10-12.7
              zypper-lifecycle-plugin:
                  ----------
                  new:
                      0.5-1.5
                  old:
              zypper-log:
                  ----------
                  new:
                      1.13.21-18.3.3
                  old:
              zypper-migration-plugin:
                  ----------
                  new:
                      0.10-12.4
                  old:
----------
          ID: test-patch-already-installed
    Function: pkg.patch_installed
      Result: True
     Comment: Related packages are already installed
     Started: 18:02:29.292585
    Duration: 994.88 ms
     Changes:   
----------
          ID: test-unexisting-patch
    Function: pkg.patch_installed
      Result: False
     Comment: An error was encountered while downloading package(s): Advisory id "SUSE-SLE-SERVER-12-SP2-2017-XXX" not found
     Started: 18:02:30.291400
    Duration: 470.47 ms
     Changes:   
----------
          ID: zip
    Function: pkg.downloaded
      Result: True
     Comment: 
     Started: 18:02:30.765749
    Duration: 399.902 ms
     Changes:   
              ----------
              zip:
                  ----------
                  new:
                      ----------
                      3.0-15.18:
                          /var/cache/zypp/packages/SLE-12-SP2-x86_64-Pool/x86_64/zip-3.0-15.18.x86_64.rpm
                  old:
----------
          ID: test-pkg-downloaded
    Function: pkg.downloaded
      Result: True
     Comment: 
     Started: 18:02:31.388246
    Duration: 372.078 ms
     Changes:   
              ----------
              whois:
                  ----------
                  new:
                      ----------
                      5.1.1-1.17:
                          /var/cache/zypp/packages/SLE-12-SP2-x86_64-Pool/x86_64/whois-5.1.1-1.17.x86_64.rpm
                  old:

Summary for local
------------
Succeeded: 5 (changed=4)
Failed:    1
------------
Total states run:     6
Total run time:  39.810 s
```

No unit tests has been added yet. I'll add them once I get some feedback about the current changes :smile: 

Do you think this PR is now in a better shape and can go in? Thanks!

cachedout commented:
@terminalmage Could you please stop by here when you get a minute to review the above changes? Thanks.

terminalmage commented on salt/modules/yumpkg.py:
```diff
@@ -2885,3 +2922,60 @@ def diff(*paths):
     return ret
+
+
+def _get_patches(installed_only=None):
```
If we're passing ``installed_only=True`` to enable it, shouldn't the default value be ``False`` and not ``None``? I get that it will work either way, but using ``False`` as a default just feels better here.

terminalmage commented on salt/modules/yumpkg.py:
```diff
@@ -907,6 +908,25 @@ def list_upgrades(refresh=True, **kwargs):
 
 
+def list_downloaded():
+    '''
```
Please add:

```
.. versionadded:: Oxygen
```

followed by a blank line, to the top of this docstring.

terminalmage commented on salt/modules/yumpkg.py:
```diff
@@ -2885,3 +2922,60 @@ def diff(*paths):
+
+
+def list_patches(refresh=False):
+    '''
```
Please add:

```
.. versionadded:: Oxygen
```

followed by a blank line, to the top of this docstring.

terminalmage commented on salt/modules/yumpkg.py:
```diff
@@ -2885,3 +2922,60 @@ def diff(*paths):
+
+
+def list_installed_patches():
+    '''
```
Please add:

```
.. versionadded:: Oxygen
```

followed by a blank line, to the top of this docstring.

terminalmage commented on salt/modules/zypper.py:
```diff
@@ -1768,6 +1779,26 @@ def download(*packages, **kwargs):
 
 
+def list_downloaded():
+    '''
```
Please add:

```
.. versionadded:: Oxygen
```

followed by a blank line, to the top of this docstring.

terminalmage commented on salt/modules/zypper.py:
```diff
@@ -1805,3 +1836,53 @@ def diff(*paths):
+
+
+def list_patches(refresh=False):
+    '''
```
Please add:

```
.. versionadded:: Oxygen
```

followed by a blank line, to the top of this docstring.

terminalmage commented on salt/modules/zypper.py:
```diff
@@ -1805,3 +1836,53 @@ def diff(*paths):
+
+
+def list_installed_patches():
+    '''
```
Please add:

```
.. versionadded:: Oxygen
```

followed by a blank line, to the top of this docstring.

terminalmage commented on salt/states/pkg.py:
```diff
@@ -1725,6 +1890,247 @@ def installed(
+               fromrepo=None,
+               ignore_epoch=None,
+               **kwargs):
+    '''
```
Please add:

```
.. versionadded:: Oxygen
```

followed by a blank line, to the top of this docstring.

terminalmage commented on salt/states/pkg.py:
```diff
@@ -1725,6 +1890,247 @@ def installed(
+
+
+def patch_installed(name, advisory_ids=None, downloadonly=None, **kwargs):
+    '''
```
Please add:

```
.. versionadded:: Oxygen
```

followed by a blank line, to the top of this docstring.

terminalmage commented on salt/states/pkg.py:
```diff
@@ -1725,6 +1890,247 @@ def installed(
+
+
+def patch_downloaded(name, advisory_ids=None, **kwargs):
+    '''
```
Please add:

```
.. versionadded:: Oxygen
```

followed by a blank line, to the top of this docstring.

terminalmage commented on salt/states/pkg.py:
```diff
@@ -1725,6 +1890,247 @@ def installed(
+          pkg.downloaded:
+            - version: 5.0.5-4.63
+            - fromrepo: "myrepository"
+    '''
```
We should gate this state so that we don't try to run it where it isn't supported.

```
if 'pkg.list_downloaded' not in __salt__:
    return {'name': name,
            'changes': {},
            'result': False,
            'comment': 'The pkg.downloaded state is not available on this platform'}
```

In fact, since we can have a lot of different reasons to exit a state early before we attempt anything, it can be helpful to make the first line of a state function a line where the return dictionary is pre-created, and then you can just set ``ret['result']`` or ``ret['comment']`` before doing a ``return ret``. This keeps you from needing to define a separate return dictionary for each case where you would return from the state.

terminalmage commented on salt/states/pkg.py:
```diff
@@ -1725,6 +1890,247 @@ def installed(
+              - SUSE-SLE-SERVER-12-SP2-2017-185
+              - SUSE-SLE-SERVER-12-SP2-2017-150
+              - SUSE-SLE-SERVER-12-SP2-2017-120
+    '''
```
Same as with ``pkg.downloaded``, let's gate this state by checking for ``pkg.list_patches`` in the ``__salt__`` dunder.

terminalmage commented on salt/states/pkg.py:
```diff
@@ -1725,6 +1890,247 @@ def installed(
+              - SUSE-SLE-SERVER-12-SP2-2017-185
+              - SUSE-SLE-SERVER-12-SP2-2017-150
+              - SUSE-SLE-SERVER-12-SP2-2017-120
+    '''
```
Like the others, let's gate this to prevent it from being used on platforms that don't support it.

terminalmage reviewed (changes requested):
This is shaping up nicely. I've added some things we should get changed. A lot of them are little notices that can be added to help flag when the new functionality was added in our documentation. That way someone a year from now doesn't look at the docs and expect that these new states/functions are available in say, the 2016.11 release cycle.

meaksh committed:
Adding versionadded flags

meaksh committed:
Some refactoring and minor fixes

meaksh commented:
@terminalmage - Thank you for your reviewing this. I've pushed new commits to address your feedback. I think everything is fixed now.

I'll add in a moment some unit tests to cover this.

meaksh committed:
Adding more unit tests for Zypper module

meaksh committed:
Pylint fix

terminalmage commented:
@meaksh I see the tests are running right now, did you push again recently? The changes look good, assuming we have no breakage in the tests this is ready for merge, in my opinion.

Thanks for all the hard work here!

meaksh commented:
@terminalmage - Great, thanks for the approval! I just pushed pylint fixes about an hour ago, maybe tests are still running. Let's see if there's nothing unexpected.

isbm commented:
@cachedout Can this also go to the Nitrogen? We had this back ported, if you like we can PR it. :wink:
